webapp/services.py

```diff
@@ -103,13 +115,45 @@ def create_asset(
         file_manager.create(file_content, url_path)
 
         # Save file info in Postgres
-        asset = Asset(file_path=url_path, data=data, created=datetime.utcnow())
+        asset = Asset(
+            file_path=url_path, data=data, tags=[], created=datetime.utcnow()
+        )
+        tags = self.create_tags_if_not_exist(tags)
+        asset.tags = tags
         db_session.add(asset)
         db_session.commit()
-
         return asset
 
-    def update_asset(self, file_path, data):
+    def create_tags_if_not_exist(self, tag_names):
+        """
+        Create the given tag name if it's new and return the
+        object from the database
+        """
+        tag_names = list(
+            set(map(lambda name: self.normalize_tag_name(name), tag_names))
+        )
+        tag_names = list(filter(lambda n: len(n) > 0, tag_names))
+        existing_tags = (
+            db_session.query(Tag).filter(Tag.name.in_(tag_names)).all()
+        )
+        tag_names_to_create = filter(
```

Comment:
What do you think about writing the bellow as:

```python3
existing_tag_names = [tag.name for tag in existing_tags]
tag_names_to_create = [name for name in tag_names if name not in existing_tag_names]
```

Reply:
I was concerned about loading the whole tags list in the memory

Reply:
The readability question aside for a second, I think we generally care more about computational complexity (performance) than spacial complexity (resources).

I *think* João's version is more performant as it avoids looking up a property from all the objects in the `existing_tags` list every time, instead building up a list of scalar strings to iterate through, which will be quicker.

It would be even quicker if that was a set (`existing_tag_names` = set(...`), as looking up items in a set is O(1) instead of O(n).

Overall I would agree with João that unless this list of tags is likely to grow into the hundreds, readability is a more important concern than either types of performance. But if we use João's version with the `set`, I think we get the best of both worlds.